File: 14-mobile-first-architecture/implementation/native/biometric-auth.kt
// Companion code for "The Backend of Luck" - Chapter 14, Mobile-First Architecture for iGaming.
// https://thebackendofluck.com | https://github.com/thebackendofluck/book
// SPDX-License-Identifier: Apache-2.0
//
// FOR TESTING AND EVALUATION ONLY. NOT FOR PRODUCTION USE.
// Published to demonstrate the patterns explained in the book. This code is
// not certified for real-money gaming: operating a gambling platform requires
// your own licence, independent test-lab certification (GLI, eCOGRA or
// equivalent) and regulator approval.

package com.acme.casino.auth

import android.content.Context
import android.os.Build
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import android.util.Log
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.IvParameterSpec

/**
 * Casino Biometric Authentication Manager (Android)
 *
 * Handles fingerprint and face authentication for:
 * - App login (session resume)
 * - Transaction confirmation (deposits, withdrawals)
 * - Responsible gaming settings changes
 *
 * Regulatory notes:
 * - Biometric auth is a second factor, not a replacement for password/KYC
 * - Session tokens stored in Android Keystore, encrypted with biometric-bound key
 * - Fallback to PIN/password always available (regulatory requirement)
 * - Auto-logout after configurable inactivity period
 */
class CasinoBiometricManager(
    private val context: Context,
    private val activity: FragmentActivity
) {
    companion object {
        private const val TAG = "CasinoBiometric"
        private const val KEYSTORE_PROVIDER = "AndroidKeyStore"
        private const val KEY_ALIAS_LOGIN = "casino_biometric_login_key"
        private const val KEY_ALIAS_TRANSACTION = "casino_biometric_txn_key"
        private const val PREFS_NAME = "casino_biometric_prefs"
        private const val PREF_ENCRYPTED_TOKEN = "encrypted_session_token"
        private const val PREF_TOKEN_IV = "session_token_iv"
        private const val PREF_BIOMETRIC_ENABLED = "biometric_enabled"
        private const val TRANSACTION_KEY_VALIDITY_SECONDS = 30
    }

    /**
     * Result callback for biometric operations
     */
    interface BiometricCallback {
        fun onSuccess(decryptedData: String?)
        fun onError(errorCode: Int, message: String)
        fun onCancelled()
    }

    /**
     * Check device biometric capabilities
     */
    fun checkBiometricAvailability(): BiometricStatus {
        val biometricManager = BiometricManager.from(context)
        return when (biometricManager.canAuthenticate(
            BiometricManager.Authenticators.BIOMETRIC_STRONG
        )) {
            BiometricManager.BIOMETRIC_SUCCESS ->
                BiometricStatus.AVAILABLE

            BiometricManager.BIOMETRIC_ERROR_NO_HARDWARE ->
                BiometricStatus.NO_HARDWARE

            BiometricManager.BIOMETRIC_ERROR_HW_UNAVAILABLE ->
                BiometricStatus.HARDWARE_UNAVAILABLE

            BiometricManager.BIOMETRIC_ERROR_NONE_ENROLLED ->
                BiometricStatus.NOT_ENROLLED

            BiometricManager.BIOMETRIC_ERROR_SECURITY_UPDATE_REQUIRED ->
                BiometricStatus.SECURITY_UPDATE_REQUIRED

            else -> BiometricStatus.UNKNOWN_ERROR
        }
    }

    /**
     * Enable biometric login by encrypting session token with biometric-bound key
     */
    fun enableBiometricLogin(sessionToken: String, callback: BiometricCallback) {
        if (checkBiometricAvailability() != BiometricStatus.AVAILABLE) {
            callback.onError(-1, "Biometric authentication not available on this device")
            return
        }

        try {
            val secretKey = getOrCreateKey(KEY_ALIAS_LOGIN, requireBiometric = true)
            val cipher = getCipher()
            cipher.init(Cipher.ENCRYPT_MODE, secretKey)

            val promptInfo = BiometricPrompt.PromptInfo.Builder()
                .setTitle("Enable Biometric Login")
                .setSubtitle("Authenticate to enable quick login")
                .setDescription("Your session will be securely stored and protected by your biometric data.")
                .setNegativeButtonText("Cancel")
                .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
                .setConfirmationRequired(true)
                .build()

            val biometricPrompt = createBiometricPrompt(
                onSuccess = { result ->
                    try {
                        val encryptedBytes = result.cryptoObject?.cipher?.doFinal(
                            sessionToken.toByteArray(Charsets.UTF_8)
                        )
                        if (encryptedBytes != null) {
                            storeEncryptedToken(
                                Base64.encodeToString(encryptedBytes, Base64.NO_WRAP),
                                Base64.encodeToString(cipher.iv, Base64.NO_WRAP)
                            )
                            setBiometricEnabled(true)
                            callback.onSuccess(null)
                            Log.i(TAG, "Biometric login enabled successfully")
                        } else {
                            callback.onError(-2, "Encryption failed")
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Failed to encrypt session token", e)
                        callback.onError(-3, "Encryption error: ${e.message}")
                    }
                },
                onError = { errorCode, message -> callback.onError(errorCode, message) },
                onCancelled = { callback.onCancelled() }
            )

            biometricPrompt.authenticate(
                promptInfo,
                BiometricPrompt.CryptoObject(cipher)
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize biometric login", e)
            callback.onError(-4, "Initialization error: ${e.message}")
        }
    }

    /**
     * Authenticate with biometrics to retrieve session token
     */
    fun authenticateForLogin(callback: BiometricCallback) {
        if (!isBiometricEnabled()) {
            callback.onError(-1, "Biometric login is not enabled")
            return
        }

        val encryptedToken = getEncryptedToken()
        val iv = getTokenIV()
        if (encryptedToken == null || iv == null) {
            callback.onError(-2, "No stored credentials found")
            disableBiometricLogin()
            return
        }

        try {
            val secretKey = getOrCreateKey(KEY_ALIAS_LOGIN, requireBiometric = true)
            val cipher = getCipher()
            cipher.init(
                Cipher.DECRYPT_MODE,
                secretKey,
                IvParameterSpec(Base64.decode(iv, Base64.NO_WRAP))
            )

            val promptInfo = BiometricPrompt.PromptInfo.Builder()
                .setTitle("Acme Casino Login")
                .setSubtitle("Authenticate to continue")
                .setNegativeButtonText("Use password instead")
                .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
                .setConfirmationRequired(false) // Faster login experience
                .build()

            val biometricPrompt = createBiometricPrompt(
                onSuccess = { result ->
                    try {
                        val decryptedBytes = result.cryptoObject?.cipher?.doFinal(
                            Base64.decode(encryptedToken, Base64.NO_WRAP)
                        )
                        if (decryptedBytes != null) {
                            val sessionToken = String(decryptedBytes, Charsets.UTF_8)
                            callback.onSuccess(sessionToken)
                            Log.i(TAG, "Biometric login successful")
                        } else {
                            callback.onError(-3, "Decryption returned null")
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Failed to decrypt session token", e)
                        // Key was invalidated (new biometric enrolled)
                        disableBiometricLogin()
                        callback.onError(-4, "Biometric data changed. Please login with your password and re-enable biometric login.")
                    }
                },
                onError = { errorCode, message -> callback.onError(errorCode, message) },
                onCancelled = { callback.onCancelled() }
            )

            biometricPrompt.authenticate(
                promptInfo,
                BiometricPrompt.CryptoObject(cipher)
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize biometric authentication", e)
            disableBiometricLogin()
            callback.onError(-5, "Authentication initialization failed. Please login with your password.")
        }
    }

    /**
     * Authenticate for high-value transaction confirmation
     * Uses a separate key with shorter validity window
     */
    fun authenticateForTransaction(
        transactionDescription: String,
        amount: String,
        callback: BiometricCallback
    ) {
        try {
            val secretKey = getOrCreateKey(
                KEY_ALIAS_TRANSACTION,
                requireBiometric = true,
                validitySeconds = TRANSACTION_KEY_VALIDITY_SECONDS
            )
            val cipher = getCipher()
            cipher.init(Cipher.ENCRYPT_MODE, secretKey)

            val promptInfo = BiometricPrompt.PromptInfo.Builder()
                .setTitle("Confirm Transaction")
                .setSubtitle(transactionDescription)
                .setDescription("Amount: $amount\n\nAuthenticate to confirm this transaction.")
                .setNegativeButtonText("Cancel")
                .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
                .setConfirmationRequired(true)
                .build()

            val biometricPrompt = createBiometricPrompt(
                onSuccess = { result ->
                    // Generate a one-time transaction token
                    val timestamp = System.currentTimeMillis()
                    val txnPayload = "$transactionDescription|$amount|$timestamp"
                    val encryptedBytes = result.cryptoObject?.cipher?.doFinal(
                        txnPayload.toByteArray(Charsets.UTF_8)
                    )
                    if (encryptedBytes != null) {
                        val txnToken = Base64.encodeToString(encryptedBytes, Base64.NO_WRAP)
                        callback.onSuccess(txnToken)
                        Log.i(TAG, "Transaction biometric confirmation successful")
                    } else {
                        callback.onError(-1, "Transaction confirmation failed")
                    }
                },
                onError = { errorCode, message -> callback.onError(errorCode, message) },
                onCancelled = { callback.onCancelled() }
            )

            biometricPrompt.authenticate(
                promptInfo,
                BiometricPrompt.CryptoObject(cipher)
            )
        } catch (e: Exception) {
            Log.e(TAG, "Transaction authentication failed", e)
            callback.onError(-2, "Please use your PIN or password to confirm this transaction.")
        }
    }

    /**
     * Disable biometric login and clear stored credentials
     */
    fun disableBiometricLogin() {
        try {
            val keyStore = KeyStore.getInstance(KEYSTORE_PROVIDER)
            keyStore.load(null)
            if (keyStore.containsAlias(KEY_ALIAS_LOGIN)) {
                keyStore.deleteEntry(KEY_ALIAS_LOGIN)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to delete biometric key", e)
        }

        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit()
            .remove(PREF_ENCRYPTED_TOKEN)
            .remove(PREF_TOKEN_IV)
            .putBoolean(PREF_BIOMETRIC_ENABLED, false)
            .apply()

        Log.i(TAG, "Biometric login disabled and credentials cleared")
    }

    // ── Private helpers ──────────────────────────────────

    private fun getOrCreateKey(
        alias: String,
        requireBiometric: Boolean,
        validitySeconds: Int = -1
    ): SecretKey {
        val keyStore = KeyStore.getInstance(KEYSTORE_PROVIDER)
        keyStore.load(null)

        keyStore.getKey(alias, null)?.let { return it as SecretKey }

        val keyGenSpec = KeyGenParameterSpec.Builder(
            alias,
            KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
        )
            .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
            .setUserAuthenticationRequired(requireBiometric)
            .setInvalidatedByBiometricEnrollment(true)
            .apply {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    setUserAuthenticationParameters(
                        if (validitySeconds > 0) validitySeconds else 0,
                        KeyProperties.AUTH_BIOMETRIC_STRONG
                    )
                } else {
                    @Suppress("DEPRECATION")
                    setUserAuthenticationValidityDurationSeconds(
                        if (validitySeconds > 0) validitySeconds else -1
                    )
                }
            }
            .build()

        val keyGenerator = KeyGenerator.getInstance(
            KeyProperties.KEY_ALGORITHM_AES,
            KEYSTORE_PROVIDER
        )
        keyGenerator.init(keyGenSpec)
        return keyGenerator.generateKey()
    }

    private fun getCipher(): Cipher {
        return Cipher.getInstance(
            "${KeyProperties.KEY_ALGORITHM_AES}/${KeyProperties.BLOCK_MODE_CBC}/${KeyProperties.ENCRYPTION_PADDING_PKCS7}"
        )
    }

    private fun createBiometricPrompt(
        onSuccess: (BiometricPrompt.AuthenticationResult) -> Unit,
        onError: (Int, String) -> Unit,
        onCancelled: () -> Unit
    ): BiometricPrompt {
        val executor = ContextCompat.getMainExecutor(context)

        return BiometricPrompt(activity, executor,
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    super.onAuthenticationSucceeded(result)
                    onSuccess(result)
                }

                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    super.onAuthenticationError(errorCode, errString)
                    if (errorCode == BiometricPrompt.ERROR_USER_CANCELED ||
                        errorCode == BiometricPrompt.ERROR_NEGATIVE_BUTTON
                    ) {
                        onCancelled()
                    } else {
                        onError(errorCode, errString.toString())
                    }
                }

                override fun onAuthenticationFailed() {
                    super.onAuthenticationFailed()
                    Log.w(TAG, "Biometric authentication attempt failed (not recognized)")
                    // Do not call onError — the system shows "not recognized" and lets user retry
                }
            }
        )
    }

    private fun storeEncryptedToken(encryptedToken: String, iv: String) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putString(PREF_ENCRYPTED_TOKEN, encryptedToken)
            .putString(PREF_TOKEN_IV, iv)
            .apply()
    }

    private fun getEncryptedToken(): String? =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getString(PREF_ENCRYPTED_TOKEN, null)

    private fun getTokenIV(): String? =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getString(PREF_TOKEN_IV, null)

    private fun isBiometricEnabled(): Boolean =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getBoolean(PREF_BIOMETRIC_ENABLED, false)

    private fun setBiometricEnabled(enabled: Boolean) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(PREF_BIOMETRIC_ENABLED, enabled)
            .apply()
    }
}

/**
 * Biometric availability status
 */
enum class BiometricStatus {
    AVAILABLE,
    NO_HARDWARE,
    HARDWARE_UNAVAILABLE,
    NOT_ENROLLED,
    SECURITY_UPDATE_REQUIRED,
    UNKNOWN_ERROR
}

File: 14-mobile-first-architecture/implementation/native/biometric-auth.swift
import LocalAuthentication
import Security
import CryptoKit
import Foundation

// MARK: - Casino Biometric Authentication Manager (iOS)
///
/// Handles Face ID and Touch ID authentication for:
/// - App login (session resume with Keychain-stored tokens)
/// - Transaction confirmation (deposits, withdrawals)
/// - Responsible gaming settings changes
///
/// Regulatory notes:
/// - Biometric auth supplements (never replaces) full KYC verification
/// - Session tokens stored in iOS Keychain with biometric access control
/// - Fallback to passcode always available (regulatory requirement)
/// - Automatic session invalidation on biometric enrollment changes

final class CasinoBiometricManager {

    // MARK: - Types

    enum BiometricType {
        case faceID
        case touchID
        case none
    }

    enum BiometricStatus {
        case available(BiometricType)
        case notAvailable(String)
        case notEnrolled
        case lockedOut
    }

    enum BiometricError: Error, LocalizedError {
        case notAvailable
        case notEnrolled
        case authenticationFailed
        case cancelled
        case keychainError(OSStatus)
        case biometricChanged
        case tokenNotFound
        case encryptionFailed

        var errorDescription: String? {
            switch self {
            case .notAvailable: return "Biometric authentication is not available on this device."
            case .notEnrolled: return "No biometric data enrolled. Please set up Face ID or Touch ID in Settings."
            case .authenticationFailed: return "Biometric authentication failed."
            case .cancelled: return "Authentication was cancelled."
            case .keychainError(let status): return "Secure storage error: \(status)"
            case .biometricChanged: return "Biometric data has changed. Please log in with your password and re-enable biometric login."
            case .tokenNotFound: return "No stored credentials found. Please log in with your password."
            case .encryptionFailed: return "Failed to secure your credentials."
            }
        }
    }

    struct TransactionConfirmation {
        let description: String
        let amount: String
        let currency: String
        let timestamp: Date
        let confirmed: Bool
    }

    // MARK: - Constants

    private enum Constants {
        static let serviceName = "com.acme.casino"
        static let sessionTokenKey = "biometric_session_token"
        static let refreshTokenKey = "biometric_refresh_token"
        static let biometricEnabledKey = "biometric_login_enabled"
    }

    // MARK: - Properties

    private let context = LAContext()

    // MARK: - Public Methods

    /// Check biometric availability and type
    func checkAvailability() -> BiometricStatus {
        var error: NSError?
        let canEvaluate = context.canEvaluatePolicy(
            .deviceOwnerAuthenticationWithBiometrics,
            error: &error
        )

        if canEvaluate {
            let type: BiometricType
            switch context.biometryType {
            case .faceID: type = .faceID
            case .touchID: type = .touchID
            default: type = .none
            }
            return .available(type)
        }

        guard let laError = error as? LAError else {
            return .notAvailable("Unknown error")
        }

        switch laError.code {
        case .biometryNotEnrolled:
            return .notEnrolled
        case .biometryLockout:
            return .lockedOut
        default:
            return .notAvailable(laError.localizedDescription)
        }
    }

    /// Enable biometric login by storing session token in Keychain with biometric protection
    func enableBiometricLogin(
        sessionToken: String,
        refreshToken: String
    ) async throws {
        guard case .available = checkAvailability() else {
            throw BiometricError.notAvailable
        }

        // Authenticate first to confirm user intent
        let authContext = LAContext()
        authContext.localizedCancelTitle = "Cancel"
        authContext.localizedReason = "Enable quick login with biometrics"

        do {
            let success = try await authContext.evaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                localizedReason: "Enable biometric login for Acme Casino"
            )

            guard success else {
                throw BiometricError.authenticationFailed
            }
        } catch let error as LAError {
            if error.code == .userCancel || error.code == .appCancel {
                throw BiometricError.cancelled
            }
            throw BiometricError.authenticationFailed
        }

        // Store tokens in Keychain with biometric access control
        try storeInKeychain(
            key: Constants.sessionTokenKey,
            data: Data(sessionToken.utf8),
            requireBiometric: true
        )

        try storeInKeychain(
            key: Constants.refreshTokenKey,
            data: Data(refreshToken.utf8),
            requireBiometric: true
        )

        UserDefaults.standard.set(true, forKey: Constants.biometricEnabledKey)
    }

    /// Authenticate with Face ID / Touch ID and retrieve session token
    func authenticateForLogin() async throws -> (sessionToken: String, refreshToken: String) {
        guard UserDefaults.standard.bool(forKey: Constants.biometricEnabledKey) else {
            throw BiometricError.tokenNotFound
        }

        guard case .available = checkAvailability() else {
            throw BiometricError.notAvailable
        }

        // Retrieve from Keychain (triggers biometric prompt automatically)
        guard let sessionData = try retrieveFromKeychain(key: Constants.sessionTokenKey),
              let refreshData = try retrieveFromKeychain(key: Constants.refreshTokenKey),
              let sessionToken = String(data: sessionData, encoding: .utf8),
              let refreshToken = String(data: refreshData, encoding: .utf8) else {
            disableBiometricLogin()
            throw BiometricError.tokenNotFound
        }

        return (sessionToken, refreshToken)
    }

    /// Authenticate for high-value transaction confirmation
    func authenticateForTransaction(
        description: String,
        amount: String,
        currency: String
    ) async throws -> TransactionConfirmation {
        guard case .available(let biometricType) = checkAvailability() else {
            throw BiometricError.notAvailable
        }

        let authContext = LAContext()
        authContext.localizedCancelTitle = "Cancel"

        // Require fresh biometric — no reuse of previous authentication
        authContext.touchIDAuthenticationAllowableReuseDuration = 0

        let reason: String
        let biometricName = biometricType == .faceID ? "Face ID" : "Touch ID"
        reason = "Confirm \(description) of \(amount) \(currency) with \(biometricName)"

        do {
            let success = try await authContext.evaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                localizedReason: reason
            )

            return TransactionConfirmation(
                description: description,
                amount: amount,
                currency: currency,
                timestamp: Date(),
                confirmed: success
            )
        } catch let error as LAError {
            switch error.code {
            case .userCancel, .appCancel:
                throw BiometricError.cancelled
            case .biometryLockout:
                // Fall back to device passcode
                return try await authenticateWithPasscode(
                    description: description,
                    amount: amount,
                    currency: currency
                )
            default:
                throw BiometricError.authenticationFailed
            }
        }
    }

    /// Authenticate for responsible gaming settings changes
    func authenticateForResponsibleGaming(
        action: String
    ) async throws -> Bool {
        let authContext = LAContext()
        authContext.touchIDAuthenticationAllowableReuseDuration = 0

        let reason = "Confirm responsible gaming change: \(action)"

        do {
            return try await authContext.evaluatePolicy(
                .deviceOwnerAuthentication, // Allow passcode fallback for RG changes
                localizedReason: reason
            )
        } catch {
            throw BiometricError.authenticationFailed
        }
    }

    /// Disable biometric login and clear stored credentials
    func disableBiometricLogin() {
        deleteFromKeychain(key: Constants.sessionTokenKey)
        deleteFromKeychain(key: Constants.refreshTokenKey)
        UserDefaults.standard.set(false, forKey: Constants.biometricEnabledKey)
    }

    /// Check if biometric login is currently enabled
    var isBiometricEnabled: Bool {
        UserDefaults.standard.bool(forKey: Constants.biometricEnabledKey)
    }

    // MARK: - Keychain Operations

    private func storeInKeychain(
        key: String,
        data: Data,
        requireBiometric: Bool
    ) throws {
        // Delete existing entry first
        deleteFromKeychain(key: key)

        // Create access control
        var accessControlFlags: SecAccessControlCreateFlags = [.privateKeyUsage]
        if requireBiometric {
            // .biometryCurrentSet invalidates when biometrics change
            accessControlFlags = [.biometryCurrentSet]
        }

        guard let accessControl = SecAccessControlCreateWithFlags(
            nil,
            kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
            accessControlFlags,
            nil
        ) else {
            throw BiometricError.encryptionFailed
        }

        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Constants.serviceName,
            kSecAttrAccount as String: key,
            kSecValueData as String: data,
            kSecAttrAccessControl as String: accessControl,
            kSecAttrSynchronizable as String: false, // Never sync biometric credentials to iCloud
        ]

        let status = SecItemAdd(query as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw BiometricError.keychainError(status)
        }
    }

    private func retrieveFromKeychain(key: String) throws -> Data? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Constants.serviceName,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
            // This will trigger biometric prompt if access control requires it
            kSecUseOperationPrompt as String: "Access your Acme Casino account",
        ]

        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        switch status {
        case errSecSuccess:
            return result as? Data
        case errSecItemNotFound:
            return nil
        case errSecUserCanceled:
            throw BiometricError.cancelled
        case errSecAuthFailed:
            throw BiometricError.authenticationFailed
        default:
            // -25293 (errSecBiometryChanged) means biometrics were re-enrolled
            if status == -25293 {
                throw BiometricError.biometricChanged
            }
            throw BiometricError.keychainError(status)
        }
    }

    private func deleteFromKeychain(key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Constants.serviceName,
            kSecAttrAccount as String: key,
        ]
        SecItemDelete(query as CFDictionary)
    }

    // MARK: - Fallback

    private func authenticateWithPasscode(
        description: String,
        amount: String,
        currency: String
    ) async throws -> TransactionConfirmation {
        let authContext = LAContext()
        let reason = "Confirm \(description) of \(amount) \(currency) with your device passcode"

        let success = try await authContext.evaluatePolicy(
            .deviceOwnerAuthentication, // Allows passcode fallback
            localizedReason: reason
        )

        return TransactionConfirmation(
            description: description,
            amount: amount,
            currency: currency,
            timestamp: Date(),
            confirmed: success
        )
    }
}

// MARK: - Usage Examples

/*
 // In your ViewController or SwiftUI View:

 let biometricManager = CasinoBiometricManager()

 // Check availability
 switch biometricManager.checkAvailability() {
 case .available(.faceID):
     showFaceIDToggle()
 case .available(.touchID):
     showTouchIDToggle()
 case .notEnrolled:
     showSetupBiometricPrompt()
 case .lockedOut:
     showPasscodeLoginOnly()
 case .notAvailable(let reason):
     hideBiometricOption()
 }

 // Enable biometric login after successful password login
 Task {
     do {
         try await biometricManager.enableBiometricLogin(
             sessionToken: loginResponse.sessionToken,
             refreshToken: loginResponse.refreshToken
         )
         showSuccess("Biometric login enabled!")
     } catch {
         showError(error.localizedDescription)
     }
 }

 // Login with biometrics
 Task {
     do {
         let tokens = try await biometricManager.authenticateForLogin()
         try await sessionManager.resume(
             sessionToken: tokens.sessionToken,
             refreshToken: tokens.refreshToken
         )
     } catch CasinoBiometricManager.BiometricError.biometricChanged {
         showAlert("Your biometric data changed. Please log in with your password.")
         biometricManager.disableBiometricLogin()
     } catch CasinoBiometricManager.BiometricError.cancelled {
         // User cancelled, do nothing
     } catch {
         showError(error.localizedDescription)
     }
 }

 // Confirm a withdrawal
 Task {
     do {
         let confirmation = try await biometricManager.authenticateForTransaction(
             description: "Withdrawal",
             amount: "500.00",
             currency: "EUR"
         )
         if confirmation.confirmed {
             try await walletService.processWithdrawal(amount: 500.0)
         }
     } catch {
         showError(error.localizedDescription)
     }
 }
 */
